src/elf.rs

```diff
@@ -719,8 +727,73 @@ impl<'a> Elf<'a> {
             self.nsections += 1;
         }
     }
+    pub fn add_extended_symtab(&mut self) -> usize {
+        // copypasta'd from add_progbits
+        let name = ".symtab_shndxr".to_string();
+        let (idx, offset) = self.new_string(name);
+        debug!(
+            "idx: {:?} @ {:#x} - new strtab offset: {:#x}",
+            idx, offset, self.sizeof_strtab
+        );
+        // the symbols section reference/index will be the current number of sections
+        let shndx = self.sections.len() + 3; // null + strtab + symtab
+        let section_symbol = SymbolBuilder::new(SymbolType::Section)
+            .section_index(shndx)
+            .create();
+
+        let section = SectionBuilder::new(((self.special_symbols.len() + self.sections.len() + self.nlocals + 1) * 4) as u64)
+            .section_type(SectionType::SymTabShndx);
+
+        let mut section = section.name_offset(offset).create(&self.ctx);
+        section.sh_link = 2;
+        // the offset is the head of how many program bits we've added
+        section.sh_offset = self.sizeof_bits as u64;
+        self.sections.insert(
+            idx,
+            SectionInfo {
+                header: section,
+                symbol: section_symbol,
+                name: idx,
+            },
+        );
+
+        // now that we inserted the final symbol, build the actual table
+        let mut data:Vec<u8> = Vec::with_capacity((self.special_symbols.len() + self.sections.len() + self.nlocals) * 4);
+        fn addpls(data:&mut Vec<u8>, mut a: u32) {
+            data.push((a & 0xff) as u8);
+            data.push(((a >> 8) & 0xff) as u8);
+            data.push(((a >> 16) & 0xff) as u8);
+            data.push(((a >> 24) & 0xff) as u8);
+        }
+
+        for symbol in &self.special_symbols {
+            addpls(&mut data, symbol.st_shndx as u32);
+        }
+        for (_id, section) in &self.sections {
+            addpls(&mut data, section.symbol.st_shndx as u32);
+        }
+        for (_id, symbol) in &self.symbols {
+            addpls(&mut data, symbol.st_shndx as u32);
```

Comment:
So if I’m right this isn’t needed (and I could be wrong) either way I believe the old symbol table for these symbols needs to be removed?

Reply:
This table doesn't replace the symtab: the symtab_shndxr is an extra table that holds a u32 for each symbol in the symbol table. If a symbol in the symtab has the special sentinel value 0xffff for the section table, it looks up the real section index in this table.
See the Oracle guide.